gno.land/pkg/gnoweb/markdown/ext_forms.go

```diff
@@ -0,0 +1,331 @@
+	if formInput.Name == "" {
+		formInput.Error = ErrFormInputMissingName
+		return parser.Continue
+	}
+
+	if formNode.InputsName[formInput.Name] {
+		formInput.Error = fmt.Errorf("%q: %w", formInput.Name, ErrFormInputNameAlreadyUsed)
+		return parser.Continue
+	}
+	formNode.InputsName[formInput.Name] = true
+
+	if formInput.Placeholder == "" {
+		formInput.Placeholder = defaultPlaceholder
+	}
+
+	if formInput.Type == "" {
+		formInput.Type = defaultInputType
+	} else if !validateInputType(formInput.Type) {
+		formInput.Error = ErrFormInvalidInputType
+		return parser.Continue
+	}
+
+	// Continue with the next line
+	return parser.Continue
+}
+
+// Close closes the block
+func (p *formParser) Close(node ast.Node, reader text.Reader, pc parser.Context) {}
+
+func (p *formParser) CanInterruptParagraph() bool { return true }
+func (p *formParser) CanAcceptIndentedLine() bool { return true }
+
+// formRenderer renders the Form node.
+// When entering the Form node, it displays the opening <form> tag
+// and when exiting (after rendering the child inputs),
+// it displays the submit button and </form>.
+type formRenderer struct{}
+
+// NewFormRenderer creates a new instance of formRenderer
+func NewFormRenderer() *formRenderer {
+	return &formRenderer{}
+}
+
+// RegisterFuncs registers the render function for the Form node
+func (r *formRenderer) RegisterFuncs(reg renderer.NodeRendererFuncRegisterer) {
+	reg.Register(KindForm, r.render)
+}
+
+// render renders the Form node
+func (r *formRenderer) render(w util.BufWriter, source []byte, node ast.Node, entering bool) (ast.WalkStatus, error) {
+	if !entering {
+		return ast.WalkContinue, nil
+	}
+
+	n, ok := node.(*FormNode)
+	if !ok {
+		return ast.WalkContinue, nil
+	}
+
+	if n.Error != nil {
+		fmt.Fprintf(w, "<!-- Error: %s -->\n", HTMLEscapeString(n.Error.Error()))
+		return ast.WalkContinue, nil
+	}
+
+	// Form action must include the full path
+	formAction := n.RealmName // start with /r/docs/markdown
+	if n.RenderPath != "" {
+		formAction += ":" + strings.TrimPrefix(n.RenderPath, "/")
+	}
+
+	// Render form opening and header
+	fmt.Fprintf(w, `<form class="gno-form" method="post" action="%s">`+"\n", HTMLEscapeString(formAction))
+	fmt.Fprintln(w, `<div class="gno-form_header">`)
+	fmt.Fprintf(w, `<span><span class="font-bold">%s</span> Form</span>`+"\n", HTMLEscapeString(n.RealmName))
+	fmt.Fprintf(w, `<span class="tooltip" data-tooltip="Processed securely by %s"><svg class="w-4 h-4"><use href="#ico-info"></use></svg></span>`+"\n", HTMLEscapeString(n.RealmName))
+	fmt.Fprintln(w, `</div>`)
+
+	for _, in := range n.Inputs {
+		if in.Error != nil {
+			fmt.Fprintf(w, "<!-- Error: %s -->\n", HTMLEscapeString(in.Error.Error()))
+			continue
+		}
+
+		// Render an input
+		fmt.Fprintf(w, `<div class="gno-form_input"><label for="%s"> %s </label>`+"\n",
+			HTMLEscapeString(in.Name),
+			HTMLEscapeString(in.Placeholder))
+		fmt.Fprintf(w, `<input type="%s" id="%s" name="%s" placeholder="%s" />`+"\n",
```

Comment:
Consider adding ` autocomplete="off"` for password inputs, just to disable autofill, which can be annoying